deepinv/physics/structured_random.py

```diff
@@ -0,0 +1,168 @@
+    change_right = math.floor(abs(input_shape[2] - output_shape[2]) / 2)
+    assert change_top + change_bottom == abs(input_shape[1] - output_shape[1])
+    assert change_left + change_right == abs(input_shape[2] - output_shape[2])
+    return torch.nn.ZeroPad2d((change_left, change_right, change_top, change_bottom))(
+        tensor
+    )
+
+
+def trimming(tensor: torch.Tensor, input_shape: tuple, output_shape: tuple):
+    r"""
+    Trimming function for undersampling in structured random phase retrieval.
+
+    :param torch.Tensor tensor: input tensor.
+    :param tuple input_shape: shape of the input tensor.
+    :param tuple output_shape: shape of the output tensor.
+
+    :return: (torch.Tensor) the trimmed tensor.
+    """
+    change_top = math.ceil(abs(input_shape[1] - output_shape[1]) / 2)
+    change_bottom = math.floor(abs(input_shape[1] - output_shape[1]) / 2)
+    change_left = math.ceil(abs(input_shape[2] - output_shape[2]) / 2)
+    change_right = math.floor(abs(input_shape[2] - output_shape[2]) / 2)
+    assert change_top + change_bottom == abs(input_shape[1] - output_shape[1])
+    assert change_left + change_right == abs(input_shape[2] - output_shape[2])
+    if change_bottom == 0:
+        tensor = tensor[..., change_top:, :]
+    else:
+        tensor = tensor[..., change_top:-change_bottom, :]
+    if change_right == 0:
+        tensor = tensor[..., change_left:]
+    else:
+        tensor = tensor[..., change_left:-change_right]
+    return tensor
+
+
+def generate_diagonal(
+    shape: tuple,
+    mode: str,
+    dtype=torch.cfloat,
+    device="cpu",
+    generator=torch.Generator("cpu"),
+):
+    r"""
+    Generate a random tensor as the diagonal matrix.
+    """
+
+    if mode == "uniform_phase":
+        diag = torch.rand(shape)
+        diag = 2 * np.pi * diag
+        diag = torch.exp(1j * diag)
+    elif mode == "rademacher":
+        diag = torch.where(
+            torch.rand(shape, device=device, generator=generator) > 0.5, -1.0, 1.0
+        )
+    else:
+        raise ValueError(f"Unsupported mode: {mode}")
+    return diag.to(device)
+
+
+class StructuredRandom(LinearPhysics):
+    r"""
+    Structured random linear operator model corresponding to the operator
+
+    .. math::
+
+        A(x) = \prod_{i=1}^N (F D_i) x,
+
+    where :math:`F` is a matrix representing a structured transform, :math:`D_i` are diagonal matrices, and :math:`N` refers to the number of layers. It is also possible to replace :math:`x` with :math:`Fx` as an additional 0.5 layer.
+
+    :param tuple input_shape: input shape. If (C, H, W), i.e., the input is a 2D signal with C channels, then zero-padding will be used for oversampling and cropping will be used for undersampling.
+    :param tuple output_shape: shape of outputs.
+    :param float n_layers: number of layers :math:`N`. If ``layers=N + 0.5``, a first :math`F` transform is included, ie :math:`A(x)=|\prod_{i=1}^N (F D_i) F x|^2`.
```

Comment:
we should give default values for ``n_layers`` (2?) ``transform_func`` (dst?) ``transform_func_inv`` (idst?) and diagonals (random +1-1s=